src/Features/LanguageServer/Protocol/Handler/Diagnostics/AbstractWorkspacePullDiagnosticsHandler.cs

```diff
@@ -0,0 +1,253 @@
+{
+    private readonly LspWorkspaceRegistrationService _workspaceRegistrationService;
+    private readonly LspWorkspaceManager _workspaceManager;
+
+    /// <summary>
+    /// Flag that represents whether the LSP view of the world has changed.
+    /// It is totally fine for this to somewhat over-report changes
+    /// as it is an optimization used to delay closing workspace diagnostic requests
+    /// until something has changed.
+    /// </summary>
+    private int _lspChanged = 0;
+
+    protected AbstractWorkspacePullDiagnosticsHandler(
+        LspWorkspaceManager workspaceManager,
+        LspWorkspaceRegistrationService registrationService,
+        IDiagnosticAnalyzerService diagnosticAnalyzerService,
+        IDiagnosticsRefresher diagnosticRefresher,
+        IGlobalOptionService globalOptions) : base(diagnosticAnalyzerService, diagnosticRefresher, globalOptions)
+    {
+        _workspaceManager = workspaceManager;
+        _workspaceRegistrationService = registrationService;
+
+        _workspaceRegistrationService.LspSolutionChanged += OnLspSolutionChanged;
+        _workspaceManager.LspTextChanged += OnLspTextChanged;
+    }
+
+    public void Dispose()
+    {
+        _workspaceManager.LspTextChanged -= OnLspTextChanged;
+        _workspaceRegistrationService.LspSolutionChanged -= OnLspSolutionChanged;
+    }
+
+    protected override async ValueTask<ImmutableArray<IDiagnosticSource>> GetOrderedDiagnosticSourcesAsync(TDiagnosticsParams diagnosticsParams, RequestContext context, CancellationToken cancellationToken)
+    {
+        // If we're being called from razor, we do not support WorkspaceDiagnostics at all.  For razor, workspace
+        // diagnostics will be handled by razor itself, which will operate by calling into Roslyn and asking for
+        // document-diagnostics instead.
+        if (context.ServerKind == WellKnownLspServerKinds.RazorLspServer)
+            return ImmutableArray<IDiagnosticSource>.Empty;
+
+        var category = GetDiagnosticCategory(diagnosticsParams);
+
+        if (category == PullDiagnosticCategories.Task)
+            return GetTaskListDiagnosticSources(context, GlobalOptions);
+
+        // if this request doesn't have a category at all (legacy behavior, assume they're asking about everything).
+        if (category == null || category == PullDiagnosticCategories.WorkspaceDocumentsAndProject)
+            return await GetDiagnosticSourcesAsync(context, GlobalOptions, cancellationToken).ConfigureAwait(false);
+
+        // if it's a category we don't recognize, return nothing.
+        return ImmutableArray<IDiagnosticSource>.Empty;
+    }
+
+    private void OnLspSolutionChanged(object? sender, WorkspaceChangeEventArgs e)
+    {
+        UpdateLspChanged();
+    }
+
+    private void OnLspTextChanged(object? sender, EventArgs e)
+    {
+        UpdateLspChanged();
+    }
+
+    private void UpdateLspChanged()
+    {
+        Interlocked.Exchange(ref _lspChanged, 1);
```

Comment:
Consider using `CancellationSeries` for this to make it much easier to avoid race conditions